diff --git a/ctgp-dif-old-2/.DS_Store b/ctgp-dif-old-2/.DS_Store
deleted file mode 100644
index 5008ddf..0000000
Binary files a/ctgp-dif-old-2/.DS_Store and /dev/null differ
diff --git a/ctgp-dif-old-2/dia.cpp b/ctgp-dif/dia.cpp
index 1ddddb5..3ee452e 100755
--- a/ctgp-dif-old-2/dia.cpp
+++ b/ctgp-dif/dia.cpp
@@ -1,5 +1,5 @@
 #include "dif.h"
-#include <bsd_string.h>
+#include <direct.h>
 #include <string.h>
 
 u8* dia_data;
@@ -74,7 +74,7 @@ int main(int argc, char** argv)
         ((DIA_Header*)dia_data)->diftab_entry_offset)];
     u32 entry_count = bswap32(((DIA_Header*)dia_data)->diftab_entry_count);
 
-    for (u32 i = 0; i < 1; ++i)
+    for (u32 i = 0; i < entry_count; ++i)
     {
         u32 infilesize = bswap32(entries[i].size);
         u8* indata = (u8*)malloc(infilesize);
@@ -87,28 +87,39 @@ int main(int argc, char** argv)
         if (int err = OpenExpFiles(pack_name, rmc_name, indata, infilesize))
         {
             printf("  failed to open expanded file %X!!\n", err);
-            return 1;
+            CloseExpFiles();
+            continue;
         }
 
-        char* out_name = (char*)malloc(65536);
-        strlcpy(out_name, pack_name, 65536);
-        strlcat(
-            out_name,
-            (const char*)&indata[bswap32(((DIF_Header*)indata)->name_offset)],
-            65536);
+        mkdir(pack_name);
+
+        char out_name[256];
+        snprintf(
+            out_name, 256, "./%s%s", pack_name,
+            (const char*)&indata[bswap32(((DIF_Header*)indata)->name_offset)]);
 
-        outfile = fopen(out_name, "wb");
-        if (!outfile)
+        if (out_name[strlen(out_name) - 1] == '/')
         {
-            printf("  failed to open output file (%s)!!\n", out_name);
-            return 1;
+            out_name[strlen(out_name) - 1] = 0;
+            printf("  creating directory (%s)\n", out_name);
+            mkdir(out_name);
+        }
+        else
+        {
+            outfile = fopen(out_name, "wb");
+            if (!outfile)
+            {
+                printf("  failed to open output file (%s)!!\n", out_name);
+                CloseExpFiles();
+                continue;
+            }
+            printf("  opened output file!\n");
+
+            ProcessDif(indata, infilesize, outfile);
+            fclose(outfile);
+
+            printf("  done!\n");
         }
-        printf("  opened output file!\n");
-
-        ProcessDif(indata, infilesize, outfile);
-        fclose(outfile);
-
-        printf("  done!\n");
     }
 
     return 0;
diff --git a/ctgp-dif-old-2/dif.cpp b/ctgp-dif/dif.cpp
index 7cbcc80..a520b23 100755
--- a/ctgp-dif-old-2/dif.cpp
+++ b/ctgp-dif/dif.cpp
@@ -1,8 +1,7 @@
 #include "dif.h"
-
-typedef unsigned char u8;
-typedef unsigned short u16;
-typedef unsigned int u32;
+#include "types.h"
+#include "yaz.h"
+#include <cstring>
 
 #define MAXFILES 15
 
@@ -15,33 +14,34 @@ u32 bswap32(u32 v)
     return (v << 24) | ((v << 8) & 0xFF0000) | ((v >> 8) & 0xFF00) | (v >> 24);
 }
 
-int OpenExpFiles(const char *pack, const char *rmc, u8 *data, u32 size)
+int OpenExpFiles(const char* pack, const char* rmc, u8* data, u32 size)
 {
+    memset(expdata, 0, sizeof(expdata));
+
     u32 expFiles = bswap32(((DIF_Header*)data)->expansion_count);
-    
+
     int i = 0;
     while (i < expFiles && i < MAXFILES)
     {
         expcursors[i + 1] = 0;
-        char *filename = (char*)malloc(65536);
-        if (((DIF_Header*)data)->exp[i].package[0] == 'R'
-                && ((DIF_Header*)data)->exp[i].package[1] == 'M'
-                && ((DIF_Header*)data)->exp[i].package[2] == 'C'
-                && !(((DIF_Header*)data)->exp[i].package[3]
-                || ((DIF_Header*)data)->exp[i].package[4]
-                || ((DIF_Header*)data)->exp[i].package[5]
-                || ((DIF_Header*)data)->exp[i].package[6]
-                || ((DIF_Header*)data)->exp[i].package[7]))
-            strlcpy(filename, rmc, 65536);
+        char filename[256];
+
+        DIF_Header* head = (DIF_Header*)data;
+        const char* prefix;
+
+        if (memcmp(head->exp[i].package, "RMC\0\0\0\0\0", 8) == 0)
+            prefix = rmc;
         else
-            strlcpy(filename, pack, 65536);
-        strlcat(filename, (char*)(data +
-                bswap32(((DIF_Header*)data)->exp[i].name_offset)), 65536);
-        
+            prefix = pack;
+
+        snprintf(
+            filename, 256, "./%s%s", prefix,
+            (char*)(data + bswap32(((DIF_Header*)data)->exp[i].name_offset)));
+
         FILE* expfile = fopen(filename, "rb");
         if (!expfile)
             return i + 1;
-        
+
         fseek(expfile, 0, SEEK_END);
         expfilesize[i] = ftell(expfile);
         expdata[i] = (u8*)malloc(expfilesize[i]);
@@ -49,32 +49,69 @@ int OpenExpFiles(const char *pack, const char *rmc, u8 *data, u32 size)
         fseek(expfile, 0, SEEK_SET);
         fread(expdata[i], 1, expfilesize[i], expfile);
         fclose(expfile);
-        
+
+        if (head->exp[i].type == TYPE_YAZ0)
+        {
+            u8* compData = expdata[i];
+
+            u32 expandSize = bswap32(*(u32*)&compData[4]);
+            printf("expandSize: %08X\n", expandSize);
+
+            u8* decompData = (u8*)malloc(expandSize);
+            assert(decompData);
+            decodeYaz0(compData + 0x10, expfilesize[i], decompData, expandSize);
+            free(expdata[i]);
+
+            expdata[i] = decompData;
+            expfilesize[i] = expandSize;
+        }
+
         printf("  finished reading expanded file %X (%s)!\n", ++i, filename);
     }
-    
+
     return 0;
 }
 
-void ProcessDif(u8 *data, u32 size, FILE *file)
+void CloseExpFiles()
 {
-    u32 dataOffset = bswap32(((DIF_Header*)data)->header_size);
-    u32 inputDataSize = bswap32(((DIF_Header*)data)->name_offset);
+    for (int i = 0; i < MAXFILES; i++)
+    {
+        if (expdata[i] != nullptr)
+            free(expdata[i]);
+    }
+}
+
+void ProcessDif(u8* data, u32 size, FILE* file)
+{
+    DIF_Header* head = (DIF_Header*)data;
+
+    u32 dataOffset = bswap32(head->header_size);
+    u32 inputDataSize = bswap32(head->name_offset);
+
+    Yaz_file_struct yaz;
+    if (head->type == TYPE_YAZ0)
+    {
+        bool ret = Yaz_open(&yaz, file);
+        assert(ret);
+    }
 
     std::vector<u8> outdata;
 
     u32 i = dataOffset;
-;
+    ;
     while (i < inputDataSize)
     {
         u32 oldlen = outdata.size();
-        
+
         u8 b = data[i++];
         u8 c = b & 15;
         int y = 0;
         u32 j = 0;
         u8 z = 0;
 
+        if (b == 0xF1)
+            break;
+
         switch (c)
         {
         case 0: {
@@ -86,7 +123,7 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
         }
         case 1: {
             if (b >> 4 == 15)
-                return;
+                break;
             y = data[i++];
             u8 s = b >> 4;
             while (s)
@@ -119,8 +156,8 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
             {
                 expcursors[z] += y + 6;
                 --z;
-                outdata.insert(outdata.end(),
-                        expdata[z] + j, expdata[z] + j + y + 6);
+                outdata.insert(outdata.end(), expdata[z] + j,
+                               expdata[z] + j + y + 6);
             }
             else
                 while (y > -6)
@@ -140,8 +177,8 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
             if (z)
             {
                 expcursors[z--] += y + 6;
-                outdata.insert(outdata.end(),
-                        expdata[z] + j, expdata[z] + j + y + 6);
+                outdata.insert(outdata.end(), expdata[z] + j,
+                               expdata[z] + j + y + 6);
             }
             else
                 while (y > -6)
@@ -155,15 +192,14 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
         }
         case 6: {
             z = b >> 4;
-            y = (data[i] + 1) * 65536
-                    + (data[i + 1] + 1) * 256 + data[i + 2];
+            y = (data[i] + 1) * 65536 + (data[i + 1] + 1) * 256 + data[i + 2];
             i += 3;
             j = expcursors[z];
             if (z)
             {
                 expcursors[z--] += y + 6;
-                outdata.insert(outdata.end(),
-                        expdata[z] + j, expdata[z] + j + y + 6);
+                outdata.insert(outdata.end(), expdata[z] + j,
+                               expdata[z] + j + y + 6);
             }
             else
                 while (y > -6)
@@ -177,15 +213,15 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
         }
         case 7: {
             z = b >> 4;
-            y = (data[i] + 1) * 1677216 + (data[i + 1] + 1) * 65536
-                    + (data[i + 2] + 1) * 256 + data[i + 3];
+            y = (data[i] + 1) * 1677216 + (data[i + 1] + 1) * 65536 +
+                (data[i + 2] + 1) * 256 + data[i + 3];
             i += 4;
             j = expcursors[z];
             if (z)
             {
                 expcursors[z--] += y + 6;
-                outdata.insert(outdata.end(),
-                        expdata[z] + j, expdata[z] + j + y + 6);
+                outdata.insert(outdata.end(), expdata[z] + j,
+                               expdata[z] + j + y + 6);
             }
             else
                 while (y > -6)
@@ -217,16 +253,15 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
         }
         case 11: {
             z = b >> 4;
-            y = (data[i] + 1) * 65536
-                    + (data[i + 1] + 1) * 256 + data[i + 2];
+            y = (data[i] + 1) * 65536 + (data[i + 1] + 1) * 256 + data[i + 2];
             i += 3;
             expcursors[z] -= (y + 1);
             break;
         }
         case 12: {
             z = b >> 4;
-            y = (data[i] + 1) * 16777216 + (data[i + 1] + 1) * 65536
-                    + (data[i + 2] + 1) * 256 + data[i + 3];
+            y = (data[i] + 1) * 16777216 + (data[i + 1] + 1) * 65536 +
+                (data[i + 2] + 1) * 256 + data[i + 3];
             i += 4;
             expcursors[z] -= (y + 1);
             break;
@@ -244,8 +279,7 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
         }
         case 15: {
             z = b >> 4;
-            expcursors[z] =
-                    data[i] * 65536 + data[i + 1] * 256 + data[i + 2];
+            expcursors[z] = data[i] * 65536 + data[i + 1] * 256 + data[i + 2];
             i += 3;
             break;
         }
@@ -255,8 +289,29 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
         }
         }
 
-        fwrite(outdata.data() + oldlen, outdata.size() - oldlen, 1, file);
-        fflush(file);
+        if (head->type == TYPE_YAZ0)
+        {
+            Yaz_write(&yaz, outdata.data() + oldlen, outdata.size() - oldlen);
+        }
+        else
+        {
+            fwrite(outdata.data() + oldlen, outdata.size() - oldlen, 1, file);
+        }
+    }
+
+    if (head->type == TYPE_YAZ0)
+    {
+        Yaz_write(&yaz, nullptr, 0);
+
+        fseek(yaz.file, 4, SEEK_SET);
+
+        u8 decompSize[4];
+        decompSize[0] = yaz.iteration >> 24;
+        decompSize[1] = yaz.iteration >> 16;
+        decompSize[2] = yaz.iteration >> 8;
+        decompSize[3] = yaz.iteration;
+
+        fwrite(decompSize, 1, 4, yaz.file);
     }
 }
 
@@ -282,20 +337,20 @@ void ProcessDif(u8 *data, u32 size, FILE *file)
     fseek(infile, 0, SEEK_SET);
     fread(indata, 1, infilesize, infile);
     fclose(infile);
-    
+
     printf("  finished reading input file!\n");
-    
+
     if (OpenExpFiles(argv[3], argv[4], indata, infilesize))
         return 1;
 
     outfile = fopen(outfilename, "wb");
     assert(outfile);
-    
+
     printf("  opened output file!\n");
 
     ProcessDif(indata, infilesize);
     fclose(outfile);
-    
+
     printf("  done!\n");
     return 0;
 }*/
diff --git a/ctgp-dif-old-2/dif.h b/ctgp-dif/dif.h
index 076892a..d37c74d 100755
--- a/ctgp-dif-old-2/dif.h
+++ b/ctgp-dif/dif.h
@@ -6,6 +6,7 @@
 #include "dif_archive.h"
 
 u32 bswap32(u32 v);
-int OpenExpFiles(const char *pack, const char *rmc, u8 *data, u32 size);
-void ProcessDif(u8 *data, u32 size, FILE *file);
-int dif_main(int argc, char **argv);
\ No newline at end of file
+int OpenExpFiles(const char* pack, const char* rmc, u8* data, u32 size);
+void CloseExpFiles();
+void ProcessDif(u8* data, u32 size, FILE* file);
+int dif_main(int argc, char** argv);
diff --git a/ctgp-dif-old-2/dif_archive.h b/ctgp-dif/dif_archive.h
index 0dd0fc8..7a08e83 100755
--- a/ctgp-dif-old-2/dif_archive.h
+++ b/ctgp-dif/dif_archive.h
@@ -1,8 +1,6 @@
 #pragma once
 
-typedef unsigned char u8;
-typedef unsigned short u16;
-typedef unsigned int u32;
+#include "types.h"
 
 #define DIA_MAGIC "DIA0"
 #define DIA_MAGIC0 'D'
@@ -44,7 +42,8 @@ typedef struct _dia_header
 
 typedef struct _dia_diftab_hdr
 {
-    char package[8]; /* sometimes "RMC", meaning this update doesn't rely on a previous version. */
+    char package[8]; /* sometimes "RMC", meaning this update doesn't rely on a
+                        previous version. */
     DIA_Version former_version; /* RMC is always 1.00.0000. */
     /* (4) zero bytes appear between the header & the actual table.
        realistically, they are probably part of the header. */
@@ -58,9 +57,9 @@ typedef struct _dia_diftab_entry
 
 typedef enum _dif_filetype : u8
 {
-    NORMAL = 0,
-    YAZ0 = 1,
-    DIRECTORY = 4
+    TYPE_NORMAL = 0,
+    TYPE_YAZ0 = 1,
+    TYPE_DIRECTORY = 4
 } FileType;
 
 typedef struct _dia_dif_exp
@@ -81,7 +80,8 @@ typedef struct _dia_dif
     u32 name_offset; /* 0x20 for directories. */
     u32 header_size; /* 0 for directories. */
     u8 unknown_14; /* always 0. */
-    u8 exists; /* 02: a file with the same name already exists. doesn't apply to SD updates. */
+    u8 exists; /* 02: a file with the same name already exists. doesn't apply to
+                  SD updates. */
     u8 unknown_16; /* always 0. */
     FileType type;
     u32 decomp_size; /* 0 for directories. */
diff --git a/ctgp-dif/types.h b/ctgp-dif/types.h
new file mode 100755
index 0000000..91ca7be
--- /dev/null
+++ b/ctgp-dif/types.h
@@ -0,0 +1,5 @@
+#pragma once
+
+typedef unsigned char u8;
+typedef unsigned short u16;
+typedef unsigned int u32;
\ No newline at end of file
diff --git a/ctgp-dif/yaz.cpp b/ctgp-dif/yaz.cpp
new file mode 100755
index 0000000..da4af02
--- /dev/null
+++ b/ctgp-dif/yaz.cpp
@@ -0,0 +1,686 @@
+#include "yaz.h"
+#include <cassert>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+
+u16 DAT_80ad1160[HASH_MAP_SIZE];
+
+u32 hash1(u32 value)
+{
+    assert(!(value & 0xff000000));
+
+    return ((value * value * 0xEF34 + value + 0xB205) >> 10) &
+           (HASH_MAP_SIZE - 1);
+}
+
+int Yaz_buffer_putcode_r(int* reent, u8* file2)
+{
+    auto file = (Yaz_file_struct*)file2;
+
+    fwrite(file->codeBuffer, file->codeBufferLocation, 1, file->file);
+    return 0;
+}
+
+#define ASSERT_FAIL(...) assert(false)
+
+static inline int Yaz_fputc_r(int* reent, int value, Yaz_file_struct* file)
+
+{
+    bool bVar1;
+    u32 uVar2;
+    u8 bVar4;
+    u32 uVar3;
+    u32 hh;
+    u8 bVar5;
+    u8 uVar6;
+    int iVar7;
+    short sVar8;
+    u32 uVar9;
+    int iVar10;
+    u32 uVar11;
+    u32 uVar12;
+    u16* puVar13;
+    u16* puVar14;
+    u16* puVar15;
+    u32 uVar16;
+    u32 uVar17;
+    u16* puVar18;
+    u16 res;
+
+    if (value == -0x1)
+    {
+        if (file->copyLocation != -0x1)
+        {
+        LAB_807eaf78:
+            uVar2 = file->copyLength;
+            if (0x3 < uVar2)
+            {
+                if (file->codeBufferIndex == '\0')
+                {
+                    file->codeBuffer[0x0] = '\0';
+                    file->codeBufferIndex = '\b';
+                    file->codeBufferLocation = '\x01';
+                }
+                do
+                {
+                    if (file->codeBufferIndex == '\0')
+                    {
+                        file->codeBufferIndex = '\b';
+                        file->codeBuffer[0x0] = '\0';
+                        file->codeBufferLocation = '\x01';
+                    }
+                    uVar17 = file->copyDistance;
+                    if ((uVar17 & 0xfffff000) != 0x0)
+                    {
+                        // WARNING: Subroutine does not return
+                        ASSERT_FAIL();
+                    }
+                    bVar5 = (u8)(uVar17 >> 0x8);
+                    if (uVar2 < 0x112)
+                    {
+                        if (0x11 < uVar2)
+                            goto LAB_807eafb4;
+                        bVar4 = file->codeBufferLocation;
+                        file->codeBuffer[bVar4] =
+                            ((char)uVar2 + -0x2) * '\x10' | bVar5;
+                        file->codeBuffer[bVar4 + 0x1 & 0xff] = (u8)uVar17;
+                        file->codeBufferLocation = bVar4 + 0x2;
+                        uVar6 = file->codeBufferIndex + 0xff;
+                        file->codeBufferIndex = uVar6;
+                    }
+                    else
+                    {
+                        uVar2 = 0x111;
+                    LAB_807eafb4:
+                        bVar4 = file->codeBufferLocation;
+                        uVar11 = (u32)bVar4;
+                        file->codeBuffer[uVar11] = bVar5;
+                        file->codeBuffer[uVar11 + 0x1 & 0xff] = (u8)uVar17;
+                        file->codeBuffer[uVar11 + 0x2 & 0xff] =
+                            (char)uVar2 + 0xee;
+                        file->codeBufferLocation = bVar4 + 0x3;
+                        uVar6 = file->codeBufferIndex + 0xff;
+                        file->codeBufferIndex = uVar6;
+                    }
+                    if ((uVar6 == '\0') &&
+                        (iVar7 = Yaz_buffer_putcode_r(reent, (u8*)file),
+                         iVar7 != 0x0))
+                    {
+                        return -0x1;
+                    }
+                    uVar2 = file->copyLength - uVar2;
+                    file->copyLength = uVar2;
+                } while (0x2 < uVar2);
+            }
+            goto LAB_807eb1a8;
+        }
+    }
+    else
+    {
+        uVar2 = file->byteShifterRemaining + 0x1;
+        iVar7 = file->iteration;
+        uVar17 = iVar7 + 0x1;
+        file->iteration = uVar17;
+        file->byteShifterRemaining = uVar2;
+        file->field5_0x1010 = value << 0x18 | file->field5_0x1010 >> 0x8;
+        if (uVar2 < 0x3)
+        {
+            return value;
+        }
+        if (0x6 < iVar7 + -0xffe)
+        {
+            puVar18 = file->hashMap;
+            uVar9 = file->field7_0x1018;
+            uVar2 = file->backBufferLocation + 0x1 & 0xfff;
+            hh = (u32)file->backBuffer[uVar2 + 0x1 & 0xfff] << 0x8 |
+                 (u32)file->backBuffer[uVar2 + 0x2 & 0xfff] << 0x10 |
+                 (u32)file->backBuffer[uVar2];
+            uVar11 = hash1(hh);
+            if (0x3fff < uVar11)
+            {
+                // WARNING: Subroutine does not return
+                ASSERT_FAIL();
+            }
+            iVar7 = 0x4001;
+            res = file->hashMap[uVar11];
+            while (true)
+            {
+                if ((res & 0x8000) == 0x0)
+                {
+                    // WARNING: Subroutine does not return
+                    ASSERT_FAIL();
+                }
+                if (((res & 0x4000) != 0x0) && (uVar2 == (res & 0x3fff)))
+                    break;
+                iVar7 = iVar7 + -0x1;
+                if (iVar7 == 0x0)
+                {
+                    // WARNING: Subroutine does not return
+                    ASSERT_FAIL();
+                }
+                uVar11 = uVar11 + 0x1 & 0x3fff;
+                res = puVar18[uVar11];
+            }
+            sVar8 = 0x1;
+            iVar7 = 0x4;
+            uVar16 = uVar2;
+            do
+            {
+                uVar12 = uVar16 + 0x2;
+                uVar3 = uVar16 + 0x3;
+                uVar16 = uVar16 + 0x1;
+                if (hh == ((u32)file->backBuffer[uVar12 & 0xfff] << 0x8 |
+                           (u32)file->backBuffer[uVar3 & 0xfff] << 0x10 |
+                           (u32)file->backBuffer[uVar16 & 0xfff]))
+                {
+                    file->hashMap[uVar11] =
+                        (short)uVar2 + sVar8 & 0xfffU | 0xc000;
+                    goto LAB_807eb550;
+                }
+                sVar8 = sVar8 + 0x1;
+                iVar7 = iVar7 + -0x1;
+            } while (iVar7 != 0x0);
+            uVar9 = uVar9 + 0x1;
+            file->hashMap[uVar11] = 0x8000;
+        LAB_807eb550:
+            file->field7_0x1018 = uVar9;
+            if (0x555 < uVar9)
+            {
+                memset(&DAT_80ad1160, 0xff, 0x8000);
+                uVar2 = 0x0;
+                puVar13 = puVar18;
+                do
+                {
+                    if (((*puVar13 & 0x8000) != 0x0) &&
+                        (uVar17 = uVar2, (*puVar13 & 0x4000) == 0x0))
+                    {
+                        do
+                        {
+                            uVar17 = uVar17 + 0x1 & 0x3fff;
+                            puVar14 = puVar13;
+                            uVar11 = uVar2;
+                        } while ((short)puVar18[uVar17] < 0x0);
+                        do
+                        {
+                            *puVar14 = 0x0;
+                            DAT_80ad1160[uVar11] = 0xffff;
+                            uVar9 = uVar17;
+                            do
+                            {
+                                do
+                                {
+                                    uVar9 = uVar9 - 0x1 & 0x3fff;
+                                    if (uVar11 == uVar9)
+                                    {
+                                        if (*puVar14 != 0x0)
+                                        {
+                                            // WARNING: Subroutine does not
+                                            // return
+                                            ASSERT_FAIL();
+                                        }
+                                        goto LAB_807eb594;
+                                    }
+                                    res = puVar18[uVar9];
+                                    puVar15 = puVar18 + uVar9;
+                                    if ((res & 0x8000) == 0x0)
+                                    {
+                                        // WARNING: Subroutine does not return
+                                        ASSERT_FAIL();
+                                    }
+                                } while ((res & 0x4000) == 0x0);
+                                hh = DAT_80ad1160[uVar9];
+                                if (hh == 0xffff)
+                                {
+                                    hh = hash1(
+                                        (u32)file->backBuffer[res + 0x1 & 0xfff]
+                                            << 0x8 |
+                                        (u32)file->backBuffer[res + 0x2 & 0xfff]
+                                            << 0x10 |
+                                        (u32)file->backBuffer[res & 0xfff]);
+                                    hh = hh & 0xffff;
+                                    DAT_80ad1160[uVar9] = (short)hh;
+                                }
+                                if (0x3fff < hh)
+                                {
+                                    // WARNING: Subroutine does not return
+                                    ASSERT_FAIL();
+                                }
+                            } while ((uVar9 - hh & 0x3fff) <
+                                     (uVar11 - hh & 0x3fff));
+                            *puVar14 = *puVar15;
+                            DAT_80ad1160[uVar11] = DAT_80ad1160[uVar9];
+                            *puVar15 = 0x8000;
+                            DAT_80ad1160[uVar9] = 0xffff;
+                            puVar14 = puVar15;
+                            uVar11 = uVar9;
+                            if ((*puVar15 & 0x4000) != 0x0)
+                            {
+                                // WARNING: Subroutine does not return
+                                ASSERT_FAIL();
+                            }
+                        } while (true);
+                    }
+                LAB_807eb594:
+                    bVar1 = uVar2 != 0x3fff;
+                    puVar13 = puVar13 + 0x1;
+                    uVar2 = uVar2 + 0x1;
+                } while (bVar1);
+                uVar17 = file->iteration;
+                file->field7_0x1018 = 0x0;
+            }
+        }
+        if (0x9 < uVar17)
+        {
+            uVar9 = file->field7_0x1018;
+            uVar2 = file->backBufferLocation - 0x3 & 0xfff;
+            uVar11 = (u32)file->backBuffer[uVar2 + 0x1 & 0xfff] << 0x8 |
+                     (u32)file->backBuffer[uVar2 + 0x2 & 0xfff] << 0x10 |
+                     (u32)file->backBuffer[uVar2];
+            iVar7 = 0x4;
+            uVar17 = uVar2;
+            do
+            {
+                if (uVar11 ==
+                    ((u32)file->backBuffer[uVar17 - 0x3 & 0xfff] << 0x8 |
+                     (u32)file->backBuffer[uVar17 - 0x2 & 0xfff] << 0x10 |
+                     (u32)file->backBuffer[uVar17 - 0x4 & 0xfff]))
+                    goto LAB_807eb78c;
+                uVar17 = uVar17 + 0x1;
+                iVar7 = iVar7 + -0x1;
+            } while (iVar7 != 0x0);
+            uVar17 = hash1(uVar11);
+            if (0x3fff < uVar17)
+            {
+                // WARNING: Subroutine does not return
+                ASSERT_FAIL();
+            }
+            res = file->hashMap[uVar17];
+            if ((res & 0x4000) != 0x0)
+            {
+                iVar7 = 0x4000;
+                while (true)
+                {
+                    uVar17 = uVar17 + 0x1 & 0x3fff;
+                    res = file->hashMap[uVar17];
+                    if ((res & 0x4000) == 0x0)
+                        break;
+                    iVar7 = iVar7 + -0x1;
+                    if (iVar7 == 0x0)
+                    {
+                        // WARNING: Subroutine does not return
+                        ASSERT_FAIL();
+                    }
+                }
+            }
+            *(u16*)(file->backBuffer + (uVar17 + 0x808) * 0x2 + -0x4) =
+                (u16)uVar2 | 0xc000;
+            uVar9 = uVar9 - (res >> 0xf);
+        LAB_807eb78c:
+            file->field7_0x1018 = uVar9;
+        }
+        iVar7 = file->copyLocation;
+        if (iVar7 != -0x1)
+        {
+            if ((file->field5_0x1010 >> 0x8 & 0xff) ==
+                (u32)file->backBuffer[iVar7])
+            {
+                uVar2 = file->backBufferLocation;
+                file->backBuffer[uVar2] = (u8)(file->field5_0x1010 >> 0x8);
+                uVar2 = uVar2 + 0x1 & 0xfff;
+                file->backBufferLocation = uVar2;
+                file->byteShifterRemaining = file->byteShifterRemaining - 0x1;
+                uVar17 = iVar7 + 0x1U & 0xfff;
+                file->copyLocation = uVar17;
+                file->copyLength = file->copyLength + 0x1;
+                if (file->copyDistance != ((uVar2 - 0x1) - uVar17 & 0xfff))
+                {
+                    // WARNING: Subroutine does not return
+                    ASSERT_FAIL();
+                }
+                return value;
+            }
+            goto LAB_807eaf78;
+        }
+    }
+    uVar2 = file->copyLength;
+LAB_807eb1a8:
+    while (uVar2 != 0x0)
+    {
+        while (true)
+        {
+            if (file->copyLocation == -0x1)
+            {
+                // WARNING: Subroutine does not return
+                ASSERT_FAIL();
+            }
+            if (file->codeBufferIndex == '\0')
+            {
+                file->codeBuffer[0x0] = '\0';
+                uVar17 = 0x7;
+                bVar5 = 0x80;
+                uVar6 = '\x02';
+                uVar11 = 0x1;
+            }
+            else
+            {
+                uVar11 = (u32)file->codeBufferLocation;
+                uVar17 = (u32)(u8)(file->codeBufferIndex - 0x1);
+                uVar6 = file->codeBufferLocation + 0x1;
+                bVar5 = (u8)(0x1 << uVar17);
+            }
+            uVar9 = file->backBufferLocation - uVar2;
+            uVar2 = uVar2 - 0x1;
+            file->codeBuffer[uVar11] = file->backBuffer[uVar9 & 0xfff];
+            file->codeBufferLocation = uVar6;
+            file->codeBufferIndex = (u8)uVar17;
+            file->copyLength = uVar2;
+            file->codeBuffer[0x0] = bVar5 | file->codeBuffer[0x0];
+            if (uVar17 != 0x0)
+                break;
+            iVar7 = Yaz_buffer_putcode_r(reent, (u8*)file);
+            if (iVar7 != 0x0)
+            {
+                return -0x1;
+            }
+            uVar2 = file->copyLength;
+            if (uVar2 == 0x0)
+                goto LAB_807eb200;
+        }
+    }
+LAB_807eb200:
+    if (value == -0x1)
+    {
+        do
+        {
+            uVar17 = file->byteShifterRemaining;
+            uVar2 = (u32)file->codeBufferIndex;
+            iVar7 = uVar17 + 0x1;
+            bVar1 = uVar2 == 0x0;
+            iVar10 = uVar17 * -0x8 + 0x20;
+            do
+            {
+                iVar7 = iVar7 + -0x1;
+                if (iVar7 == 0x0)
+                {
+                    if ((!bVar1) &&
+                        (iVar7 = Yaz_buffer_putcode_r(reent, (u8*)file),
+                         iVar7 != 0x0))
+                    {
+                        return -0x1;
+                    }
+                    if (file->field12_0x9028 == 0x0)
+                    {
+                        return 0x0;
+                    }
+                    uVar2 = fwrite(file->field13_0x902c, 0x1,
+                                   file->field12_0x9028, file->file);
+                    if (uVar2 == file->field12_0x9028)
+                    {
+                        return 0x0;
+                    }
+                    return -0x1;
+                }
+                file->backBuffer[file->backBufferLocation] =
+                    (u8)(file->field5_0x1010 >> iVar10);
+                file->backBufferLocation =
+                    file->backBufferLocation + 0x1 & 0xfff;
+                if (bVar1)
+                {
+                    file->codeBuffer[0x0] = (u8)uVar2;
+                    bVar5 = 0x80;
+                    uVar2 = 0x7;
+                    uVar6 = '\x02';
+                    uVar11 = 0x1;
+                }
+                else
+                {
+                    uVar11 = (u32)file->codeBufferLocation;
+                    uVar2 = uVar2 - 0x1 & 0xff;
+                    uVar6 = file->codeBufferLocation + 0x1;
+                    bVar5 = (u8)(0x1 << uVar2);
+                }
+                bVar1 = uVar2 == 0x0;
+                uVar17 = uVar17 - 0x1;
+                iVar10 = iVar10 + 0x8;
+                file->codeBuffer[uVar11] =
+                    (u8)(file->field5_0x1010 >>
+                         file->byteShifterRemaining * (-0x8 + 0x20));
+                file->codeBufferLocation = uVar6;
+                file->codeBufferIndex = (u8)uVar2;
+                file->codeBuffer[0x0] = bVar5 | file->codeBuffer[0x0];
+                file->byteShifterRemaining = uVar17;
+            } while (!bVar1);
+            iVar7 = Yaz_buffer_putcode_r(reent, (u8*)file);
+        } while (iVar7 == 0x0);
+    }
+    else
+    {
+        if (file->byteShifterRemaining != 0x3)
+        {
+            // WARNING: Subroutine does not return
+            ASSERT_FAIL();
+        }
+        uVar17 = file->field5_0x1010;
+        uVar2 = file->backBufferLocation;
+        uVar6 = (u8)(uVar17 >> 0x8);
+        file->backBuffer[uVar2] = uVar6;
+        uVar2 = uVar2 + 0x1 & 0xfff;
+        file->backBufferLocation = uVar2;
+        file->byteShifterRemaining = 0x2;
+        if ((u32)file->iteration < 0x3)
+        {
+        LAB_807eb248:
+            if (file->codeBufferIndex == '\0')
+            {
+                file->codeBuffer[0x1] = uVar6;
+                file->codeBufferLocation = '\x02';
+                file->codeBufferIndex = '\a';
+                file->codeBuffer[0x0] = 0x80;
+            }
+            else
+            {
+                bVar5 = file->codeBufferLocation;
+                bVar4 = file->codeBufferIndex - 0x1;
+                file->codeBuffer[bVar5] = uVar6;
+                file->codeBufferLocation = bVar5 + 0x1;
+                file->codeBufferIndex = bVar4;
+                file->codeBuffer[0x0] =
+                    (u8)(0x1 << (u32)bVar4) | file->codeBuffer[0x0];
+                if (bVar4 == 0x0)
+                {
+                    iVar7 = Yaz_buffer_putcode_r(reent, (u8*)file);
+                    if (iVar7 != 0x0)
+                    {
+                        return -0x1;
+                    }
+                    uVar2 = file->backBufferLocation;
+                    uVar17 = file->field5_0x1010;
+                }
+            }
+            file->copyLength = 0x0;
+            file->copyDistance = 0x0;
+            uVar11 = uVar2 - 0x1 & 0xfff;
+            file->copyLocation = -0x1;
+            uVar17 = uVar17 >> 0x10 & 0xff;
+            if (file->backBuffer[uVar11] == uVar17)
+            {
+                file->copyLocation = uVar11;
+            }
+            uVar2 = uVar2 - 0x2 & 0xfff;
+            if (uVar17 == file->backBuffer[uVar2])
+            {
+                file->copyLocation = uVar2;
+                file->copyDistance = 0x1;
+            }
+        }
+        else
+        {
+            uVar11 = hash1(uVar17 >> 0x8);
+            iVar7 = 0x4001;
+            if (0x3fff < uVar11)
+            {
+                // WARNING: Subroutine does not return
+                ASSERT_FAIL();
+            }
+            uVar9 = (u32)file->hashMap[uVar11];
+            if ((file->hashMap[uVar11] & 0x4000) == 0x0)
+                goto LAB_807eb408;
+            while (true)
+            {
+                if ((uVar9 & 0x8000) == 0x0)
+                {
+                    // WARNING: Subroutine does not return
+                    ASSERT_FAIL();
+                }
+                hh = uVar9 & 0xfff;
+                if (uVar17 >> 0x8 ==
+                    ((u32)file->backBuffer[uVar9 + 0x1 & 0xfff] << 0x8 |
+                     (u32)file->backBuffer[uVar9 + 0x2 & 0xfff] << 0x10 |
+                     (u32)file->backBuffer[hh]))
+                    break;
+                while (true)
+                {
+                    iVar7 = iVar7 + -0x1;
+                    if (iVar7 == 0x0)
+                    {
+                        // WARNING: Subroutine does not return
+                        ASSERT_FAIL();
+                    }
+                    uVar11 = uVar11 + 0x1 & 0x3fff;
+                    uVar9 = (u32)file->hashMap[uVar11];
+                    if ((file->hashMap[uVar11] & 0x4000) != 0x0)
+                        break;
+                LAB_807eb408:
+                    if ((uVar9 & 0x8000) == 0x0)
+                        goto LAB_807eb248;
+                }
+            }
+            uVar17 = uVar2 - 0x1 & 0xfff;
+            if (file->backBuffer[uVar17] != file->backBuffer[hh])
+            {
+                // WARNING: Subroutine does not return
+                ASSERT_FAIL();
+            }
+            if (uVar17 == hh)
+            {
+                // WARNING: Subroutine does not return
+                ASSERT_FAIL();
+            }
+            uVar17 = hh + 0x1 & 0xfff;
+            file->copyLength = 0x1;
+            file->copyLocation = uVar17;
+            file->copyDistance = (uVar2 - 0x1) - uVar17 & 0xfff;
+        }
+    }
+    return value;
+}
+
+bool Yaz_open(Yaz_file_struct* file, FILE* fptr)
+{
+    memset(file, 0, sizeof(Yaz_file_struct));
+
+    file->file = fptr;
+    if (!file->file)
+        return false;
+
+    const u8 szsHeader[] = {
+        0x59, 0x61, 0x7A, 0x30, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    };
+
+    int ret = fwrite(szsHeader, sizeof(szsHeader), 1, file->file);
+    assert(ret == 1);
+
+    file->copyLocation = -1;
+    return true;
+}
+
+bool Yaz_write(Yaz_file_struct* file, u8* data, u32 len)
+{
+    int reent;
+
+    if (data != nullptr)
+    {
+        for (u32 i = 0; i < len; i++)
+        {
+            int ret = Yaz_fputc_r(&reent, data[i], file);
+            if (ret < 0)
+                return false;
+        }
+    }
+    else
+    {
+        int ret = Yaz_fputc_r(&reent, -1, file);
+        if (ret < 0)
+            return false;
+    }
+
+    return true;
+}
+
+Ret decodeYaz0(u8* src, int srcSize, u8* dst, int uncompressedSize)
+{
+    Ret r = {0, 0};
+    // int srcPlace = 0, dstPlace = 0; //current read/write positions
+
+    u32 validBitCount = 0; // number of valid bits left in "code" byte
+    u8 currCodeByte;
+    while (r.dstPos < uncompressedSize)
+    {
+        // read new "code" byte if the current one is used up
+        if (validBitCount == 0)
+        {
+            currCodeByte = src[r.srcPos];
+            ++r.srcPos;
+            validBitCount = 8;
+        }
+
+        if ((currCodeByte & 0x80) != 0)
+        {
+            // straight copy
+            dst[r.dstPos] = src[r.srcPos];
+            r.dstPos++;
+            r.srcPos++;
+            // if(r.srcPos >= srcSize)
+            //   return r;
+        }
+        else
+        {
+            // RLE part
+            u8 byte1 = src[r.srcPos];
+            u8 byte2 = src[r.srcPos + 1];
+            r.srcPos += 2;
+            // if(r.srcPos >= srcSize)
+            //   return r;
+
+            u32 dist = ((byte1 & 0xF) << 8) | byte2;
+            u32 copySource = r.dstPos - (dist + 1);
+
+            u32 numBytes = byte1 >> 4;
+            if (numBytes == 0)
+            {
+                numBytes = src[r.srcPos] + 0x12;
+                r.srcPos++;
+                // if(r.srcPos >= srcSize)
+                //   return r;
+            }
+            else
+                numBytes += 2;
+
+            // copy run
+            for (int i = 0; i < numBytes; ++i)
+            {
+                dst[r.dstPos] = dst[copySource];
+                copySource++;
+                r.dstPos++;
+            }
+        }
+
+        // use next bit from "code" byte
+        currCodeByte <<= 1;
+        validBitCount -= 1;
+    }
+
+    return r;
+}
diff --git a/ctgp-dif/yaz.h b/ctgp-dif/yaz.h
new file mode 100755
index 0000000..ed53e06
--- /dev/null
+++ b/ctgp-dif/yaz.h
@@ -0,0 +1,38 @@
+#pragma once
+
+#include "types.h"
+#include <cstdio>
+
+#define BACK_BUF_SIZE 0x1000
+#define HASH_MAP_SIZE 0x4000
+
+struct Yaz_file_struct
+{
+    FILE* file;
+    u32 _0x4;
+    int iteration;
+    u32 backBufferLocation; // C
+    u8 backBuffer[BACK_BUF_SIZE];
+    u32 field5_0x1010;
+    u32 byteShifterRemaining;
+    u32 field7_0x1018;
+    u16 hashMap[HASH_MAP_SIZE]; // 101c
+    int copyLocation; // 901c
+    u32 copyDistance; // 9020
+    u32 copyLength;
+    u32 field12_0x9028;
+    u8 field13_0x902c[0x922C - 0x902C];
+    u8 codeBufferLocation; // 922c
+    u8 codeBufferIndex; // 922d
+    u8 codeBuffer[0x9248 - 0x922E]; // not actual size probably
+};
+
+bool Yaz_open(Yaz_file_struct* file, FILE* fptr);
+bool Yaz_write(Yaz_file_struct* file, u8* data, u32 len);
+
+struct Ret
+{
+    int srcPos, dstPos;
+};
+
+Ret decodeYaz0(u8* src, int srcSize, u8* dst, int uncompressedSize);
